### PR TITLE
Use W3C trace context headers to support OTEL trace propagation

### PR DESCRIPTION
**Pull request by swfunc**
### Adding propagation to OTEL tracing
Implements OTEL propagation https://opentelemetry.io/docs/instrumentation/js/propagation/ for API route handler traces by extracting the W3C trace context headers https://www.w3.org/TR/trace-context/ from the request and using them to create the context for the active span.

I spent some time trying to figure out how best to test this but I wasn't sure. Would appreciate a pointer to existing tests for the tracer or any advice on testing this.

I followed the doc here https://github.com/vercel/next.js/blob/canary/contributing/core/developing-using-local-app.md and was able to confirm the header values are being used to create the context through logging, but have not been able to do a more thorough test end-to-end to confirm that the generated API route spans are parented correctly. I wanted to deploy a testing app to Vercel that uses my local Next.js build so I could double check that the headers are being used in the generated traces, but I ran into some problems getting the app to build correctly with my local version of Next.js. I asked in Discord here https://discord.com/channels/752553802359505017/1007476603422527558/threads/1162417878507737228 and someone has been helping out but I still haven't yet been able to deploy it. If the reviewer has any advice on the `Error: No Next.js version could be detected in your project. Make sure `"next"` is installed in "dependencies" or "devDependencies"` error I'm getting there it would be appreciated (more context in Discord). 

*I was not able to find a checklist in the PR template for this that is a great match for this since it's not exactly a bug or a feature. Please let me know if anything should be added or expanded here

### CODE REVIEW

File: packages/next/src/server/base-server.ts [hunk 1]
```diff
@@ -711,6 +720,7 @@ export default abstract class Server<ServerOptions extends Options = Options> {
           'http.method': method,
           'http.target': req.url,
         },
+        reqHeaderTraceContext: { traceParent, traceState },
```
Comment (feedthejim) [inline packages/next/src/server/base-server.ts:723]: I don't think we need the intermediary context object. Also let's use the `traceparent` as is, without the camelcase, to follow the OTEL convention

**Comment by feedthejim** [comment]
to test it out yourself, you can run `pnpm pack` inside the `packages/next` folder and install the tarball to your project

**Comment by valkum** [comment]
As someone who maintains an internal patch for this support (hadn't had time to create an upstream PR yet): Can we get some tests as well to avoid breaking this in the future?

File: packages/next/src/server/base-server.ts (hunk 1 above)
Comment (sfirrin) [reply packages/next/src/server/base-server.ts:723]: >I don't think we need the intermediary context object

To clarify, do you mean to include the traceparent and tracestate properties in the top level of this TracerSpanOptions object?

**Comment by sfirrin** [comment]
Definitely agree @valkum, I am trying to figure out how to test this but haven't been able to find how this kind of tracer logic is tested within Next.js so would appreciate any pointer

**Comment by valkum** [comment]
I think here would be the right place: https://github.com/vercel/next.js/blob/canary/test/e2e/opentelemetry/opentelemetry.test.ts

File: packages/next/src/server/base-server.ts [hunk 2]
```diff
@@ -702,6 +702,15 @@ export default abstract class Server<ServerOptions extends Options = Options> {
   ): Promise<void> {
     await this.prepare()
     const method = req.method.toUpperCase()
+
+    // Extract the w3c trace context headers and pass them to tracer
+    const traceParentHeader = req.headers.traceparent
```
Comment (dvoytenko) [inline packages/next/src/server/base-server.ts:707]: IMHO, it'd be better to either move the propagation extraction code from below to here or to pass headers to the `trace` call. For instance, this code could execute:

```
spanContext = propagation.extract(context.active(), req.headers)
```

And pass an optional `spanContext` to the `trace()` method. This way the SDK can configure whatever propagation strategy it likes outside of Next and it'd work.

File: packages/next/src/server/base-server.ts [hunk 3]
```diff
@@ -702,6 +702,15 @@ export default abstract class Server<ServerOptions extends Options = Options> {
...
+    const traceStateHeader = req.headers.tracestate
+    const traceParent =
```
Comment (dvoytenko) [inline packages/next/src/server/base-server.ts:709]: If we extract the propagation, should we also inject it in the downstream fetch calls? Or would that be too subjective?

**Review by dvoytenko — CHANGES_REQUESTED** [review]

File: packages/next/src/server/base-server.ts (hunk 2 above)
Comment (dvoytenko) [reply packages/next/src/server/base-server.ts:707]: If we don't want to pass `spanContext` to `trace()`, there should also be a way to propagate it via OpenTelemetry's `context.with()`.

Comment (dvoytenko) [reply packages/next/src/server/base-server.ts:707]: Here's the alternative, which, IMHO, should cover the target cases better: https://github.com/vercel/next.js/pull/57084

Comment (sfirrin) [reply packages/next/src/server/base-server.ts:707]: OK thanks, I was planning to work on your feedback today but I'm happy to let you take it from here with your PR

Can you help me understand how your PR is an improvement / expand on what you mean by `This way the SDK can configure whatever propagation strategy it likes outside of Next and it'd work`?

Comment (dvoytenko) [reply packages/next/src/server/base-server.ts:707]: In a otel setup (in an application, not in Next), you could do something like this:

```
traceProvider.register({
  propagator: new XPropagator(),
})
```

E.g. you could use [`AWSXRayPropagator`](https://github.com/open-telemetry/opentelemetry-js-contrib/blob/main/propagators/opentelemetry-propagator-aws-xray/src/AWSXRayPropagator.ts) or make your own.

This propagator could work off headers or some special environment context. Next.js wouldn't need to know any of that - it'd just call a generic `propagator.extract()` and  (eventually)`propagator.inject()` and delegate this part fully to the configuration of the application or a deployment environment.

Comment (dvoytenko) [reply packages/next/src/server/base-server.ts:707]: And FWIW, your PR is super close, it just limits which propagators can be used.

Comment (sfirrin) [reply packages/next/src/server/base-server.ts:707]: I see, thanks for explaining and for opening your PR. I'll close this one out in favor of yours

**Comment by sfirrin** [comment]
Closing in favor of https://github.com/vercel/next.js/pull/57084 which does the same

**Pull request closed without merging**